### PR TITLE
Add Jersey support for Java 8 DateTimes

### PR DESCRIPTION
**Pull request by markelliot**
This change is [Reviewable](https://reviewable.io/reviews/palantir/http-remoting/399)

### CODE REVIEW

File: jersey-servers/src/main/java/com/palantir/remoting2/servers/jersey/ZonedDateTimeParamConverterProvider.java [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    @SuppressWarnings("unchecked")
+    public <T> ParamConverter<T> getConverter(final Class<T> rawType, final Type genericType,
+                                              final Annotation[] annotations) {
+        return ZonedDateTime.class.equals(rawType) ? (ParamConverter<T>) paramConverter : null;
+    }
+
+    public static final class ZonedDateTimeParamConverter implements ParamConverter<ZonedDateTime> {
+        @Override
+        public ZonedDateTime fromString(final String value) {
+            return ZonedDateTime.parse(value);
```
Comment (uschi2000) [inline jersey-servers/src/main/java/com/palantir/remoting2/servers/jersey/ZonedDateTimeParamConverterProvider.java:42]: how does this gel with the jackson way of deserializing such things?
```
                .disable(SerializationFeature.WRITE_DATES_AS_TIMESTAMPS)
                .disable(DeserializationFeature.ADJUST_DATES_TO_CONTEXT_TIME_ZONE)
```

Comment (markelliot) [reply jersey-servers/src/main/java/com/palantir/remoting2/servers/jersey/ZonedDateTimeParamConverterProvider.java:42]: it's the same. I attempted to write a JacksonParamConverterProvider that returned a ParamConverter for all types but ran into two issues:
 1. It seems like there are some types which need to be ignored
 2. Jackson cannot deserialize null or empty bodies into Optionals/instead throws on null or empty bodies

Comment (markelliot) [reply jersey-servers/src/main/java/com/palantir/remoting2/servers/jersey/ZonedDateTimeParamConverterProvider.java:42]: It's possible some more effort might yield better results, I did the quick thing, here, to unblock some stuff

Comment (uschi2000) [reply jersey-servers/src/main/java/com/palantir/remoting2/servers/jersey/ZonedDateTimeParamConverterProvider.java:42]: ok, cool. maybe add one test that checks the jaxrsclient -> jersey -> jaxrsclient serde path.

**Review by uschi2000 — COMMENTED** [review]
add one test, otherwise +1

**Review by uschi2000 — APPROVED** [review]

File: jersey-servers/src/main/java/com/palantir/remoting2/servers/jersey/ZonedDateTimeParamConverterProvider.java (hunk 1 above)
Comment (markelliot) [reply jersey-servers/src/main/java/com/palantir/remoting2/servers/jersey/ZonedDateTimeParamConverterProvider.java:42]: didn't actually do the jaxrs side of things just yet =/

**Comment by uschi2000** [comment]
oh, yeah this is the path that doesn't go through jackson, right.

**Comment by uschi2000** [comment]
do we still need this?

**Comment by markelliot** [comment]
yes

File: jersey-servers/src/test/java/com/palantir/remoting2/servers/jersey/DateTimeTest.java [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+    }
+
+    @Test
+    public void testInstantParam() {
+        Response response = target.path("instant").queryParam("value", "2017-01-02T03:04:05.06Z").request().get();
+        assertThat(response.getStatus(), is(Status.OK.getStatusCode()));
+        assertThat(response.readEntity(String.class), is("2017-01-02T03:04:05.060Z"));
+    }
+
+    public static class OptionalTestServer extends Application<Configuration> {
```
Comment (gracew) [inline jersey-servers/src/test/java/com/palantir/remoting2/servers/jersey/DateTimeTest.java:74]: probably want to rename this to DateTimeTestServer

**Pull request merged**